xds/internal/xdsclient/xds.go

```diff
@@ -284,59 +277,56 @@ func processServerSideListener(lis *v3listenerpb.Listener) (*ListenerUpdate, err
+			// directly contain the serialized bytes of the actual filter
+			// configuration, or it could be encoded as a `TypedStruct`.
+			// TODO: Add support for `TypedStruct`.
+			tc := filter.GetTypedConfig()
+
+			// The only network filter that we currently support is the v3
+			// HttpConnectionManager. So, we can directly check the type_url
+			// and unmarshal the config.
+			// TODO: Implement a registry of supported network filters (like
+			// we have for HTTP filters), when we have to support network
+			// filters other than HttpConnectionManager.
+			if tc.GetTypeUrl() != version.V3HTTPConnManagerURL {
+				return nil, fmt.Errorf("filter chain {%+v} has unsupported network filter %q in filter {%+v}", filterChain, tc.GetTypeUrl(), filter)
 			}
-			if seenNames[name] {
-				return fmt.Errorf("filter chain {%+v} has duplicate filter name %q", filterChain, name)
+			hcm := &v3httppb.HttpConnectionManager{}
+			if err := ptypes.UnmarshalAny(tc, hcm); err != nil {
+				return nil, fmt.Errorf("filter chain {%+v} failed unmarshaling of network filter {%+v}: %v", filterChain, filter, err)
 			}
-			seenNames[name] = true
-
-			// Network filters have a oneof field named `config_type` where we
-			// only support `TypedConfig` variant.
-			switch typ := filter.GetConfigType().(type) {
-			case *v3listenerpb.Filter_TypedConfig:
-				// The typed_config field has an `anypb.Any` proto which could
-				// directly contain the serialized bytes of the actual filter
-				// configuration, or it could be encoded as a `TypedStruct`.
-				// TODO: Add support for `TypedStruct`.
-				tc := filter.GetTypedConfig()
-
-				// The only network filter that we currently support is the v3
-				// HttpConnectionManager. So, we can directly check the type_url
-				// and unmarshal the config.
-				// TODO: Implement a registry of supported network filters (like
-				// we have for HTTP filters), when we have to support network
-				// filters other than HttpConnectionManager.
-				if tc.GetTypeUrl() != version.V3HTTPConnManagerURL {
-					return fmt.Errorf("filter chain {%+v} has unsupported network filter %q in filter {%+v}", filterChain, tc.GetTypeUrl(), filter)
-				}
-				hcm := &v3httppb.HttpConnectionManager{}
-				if err := ptypes.UnmarshalAny(tc, hcm); err != nil {
-					return fmt.Errorf("filter chain {%+v} failed unmarshaling of network filter {%+v}: %v", filterChain, filter, err)
-				}
-				// We currently don't support HTTP filters on the server-side.
-				// We will be adding support for it in the future. So, we want
-				// to make sure that the http_filters configuration is valid.
-				if _, err := processHTTPFilters(hcm.GetHttpFilters(), true); err != nil {
-					return err
-				}
-				seenHCM = true
-			default:
-				return fmt.Errorf("filter chain {%+v} has unsupported config_type %T in filter %s", filterChain, typ, filter.GetName())
+			var err error
+			if httpFilters, err = processHTTPFilters(hcm.GetHttpFilters(), true); err != nil {
```

Comment:
Hmm, this overrides `httpFilters`, so we only keep the last one from the loop? Is that right?

Reply:
And let's make sure the behavior is covered in some tests (multiple filters, and multipler httpfilters in each filter).

Reply:
Is it guaranteed to appear exactly once?
If it appears multiple times, is this response still valid (e.g. should we NACK it)?
If it's valid, we still need to decided whether to use the first one or the last one. The current code keeps the last one, we can make it keep the first one by `break` from the loop.

In addition, this is pretty confusing, let's also add comment in the code.

Reply:
"Additionally, the envoy.extensions.filters.network.http_connection_manager.v3.HttpConnectionManager filter, HttpConnectionManager hereafter, must be present. Any filters after HttpConnectionManager should be ignored during connection processing but must still be considered for validity." - A36-xds-for-servers.

Reply:
Here's my thought: rather than break, have a conditional which wraps this. That way, the other network filters still get validated (which in this case checks for duplicate names + whether it's HCM or not), but "are ignored during connection processing"